File: 1605. Find Valid Matrix Given Row and Column Sums/restoreMatrix.py
from typing import List

class Solution:
  def restoreMatrix(self, rowSum: List[int], colSum: List[int]) -> List[List[int]]:
    ROWS, COLS = len(rowSum), len(colSum)

    res = [[0] * COLS for _ in range(ROWS)]

    for r in range(ROWS):
      res[r][0] = rowSum[r]

    for c in range(COLS):
      currColSum = 0
      for r in range(ROWS):
        currColSum += res[r][c]

      r = 0
      while currColSum > colSum[c]:
        diff = currColSum - colSum[c]
        maxShift = min(res[r][c], diff)

        res[r][c] -= maxShift
        res[r][c + 1] += maxShift
        currColSum -= maxShift
        r += 1

    return res


rowSum = [5, 7, 10]
colSum = [8, 6, 8]
solution = Solution()
print('restoreMatrix: ', solution.restoreMatrix(rowSum, colSum))

File: 1605. Find Valid Matrix Given Row and Column Sums/restoreMatrix.js
/**
 * @param {number[]} rowSum
 * @param {number[]} colSum
 * @return {number[][]}
 */
const restoreMatrix = (rowSum, colSum) => {
  const ROWS = rowSum.length;
  const COLS = colSum.length;

  let r = 0;
  let c = 0;

  const originalMatrix = Array.from({ length: ROWS }, () => Array(COLS).fill(0));

  while (r < ROWS && c < COLS) {
    const minVal = Math.min(rowSum[r], colSum[c]);

    originalMatrix[r][c] = minVal;

    rowSum[r] -= minVal;
    colSum[c] -= minVal;

    if (colSum[c] === 0) {
      c++;
    } else {
      r++;
    }
  }

  return originalMatrix;
};

const rowSum = [5, 7, 10];
const colSum = [8, 6, 8];
console.log('restoreMatrix: ', restoreMatrix(rowSum, colSum))
